src/aiohomeconnect/client.py

```diff
@@ -69,13 +83,247 @@ async def request(self, method: str, url: str, **kwargs: Any) -> Response:
+                    )
+                    await self._message_handler(event)
+            except ReadTimeout:
+                _LOGGER.debug("Connection timed out")
+            else:
+                _LOGGER.debug("Connection closed cleanly")
+
+    async def stream_all_events(
+        self,
+        *,
+        accept_language: Language | None = None,
+    ) -> None:
+        """Stream all events from the Home Connect API."""
+        await self._stream(
+            "/homeappliances/events",
+            accept_language=accept_language,
+        )
+
+    async def stream_events(
+        self,
+        ha_id: str,
+        *,
+        accept_language: Language | None = None,
+    ) -> None:
+        """Stream events for a specific home appliance from the Home Connect API."""
+        await self._stream(
+            f"/homeappliances/{ha_id}/events",
+            accept_language=accept_language,
+        )
+
+    async def _message_handler(self, message_event: ServerSentEvent) -> None:
+        """Match event with callback for event type."""
+        event_type = EventTypes(message_event.event)
+        if event_type == EventTypes.KEEP_ALIVE:
+            return
+        ha_id = message_event.id
+        match event_type:
+            case EventTypes.STATUS | EventTypes.EVENT | EventTypes.NOTIFY:
+                try:
+                    events = ArrayOfEvents.from_json(message_event.data)
+                except JSONDecodeError as err:
+                    _LOGGER.debug("Error parsing event: %s", err)
+                except InvalidFieldValue as err:
+                    _LOGGER.debug("Error parsing event: %s", err)
+                else:
+                    await self._event_handler(events, event_type, ha_id)
+            case (
+                EventTypes.CONNECTED
+                | EventTypes.DISCONNECTED
+                | EventTypes.PAIRED
+                | EventTypes.DEPAIRED
+            ):
+                await self.home_appliances_changes_callback(event_type, ha_id)
+
+    async def _event_handler(
+        self, events: ArrayOfEvents, event_type: EventTypes, ha_id: str
+    ) -> None:
+        async def default_function(_: Event) -> None:
+            pass
+
+        for event in events.items:
+            match event_type:
+                case EventTypes.STATUS:
+                    callback = self._status_callbacks.get(ha_id, {}).get(
+                        event.key,
+                        default_function,
+                    )
+                case EventTypes.EVENT:
+                    callback = self._event_callbacks.get(ha_id, {}).get(
+                        event.key,
+                        default_function,
+                    )
+                case EventTypes.NOTIFY:
+                    callback = self._notify_callbacks.get(ha_id, {}).get(
+                        event.key,
+                        default_function,
+                    )
+            await callback(event)
```

Comment:
I don't want to handle callbacks in the library. We should just yield the events from an async iterator.

Reply:
If not all the event types match the current `Event` model we should add a base event model that match all events and new child event models as needed.

Reply:
> all the event types match the current `Event` model we should add a base event model that match all events and new child event models as needed.

So do you mean that we need something like this?
```python
@dataclass
class EventBase(DataClassJSONMixin):
    """Represent Event."""

    key: EventKey
    timestamp: int
    level: str
    handling: str
    value: str | int | float | bool


@dataclass
class EventWithName(EventBase):
    """Represent Event with name."""

    name: str


@dataclass
class EventWithUri(EventBase):
    """Represent Event with uri."""

    uri: str

@dataclass
class EventWithUriAndName(EventWithName, EventWithUri):
    """Represent Event with uri and name."""

# [etc.]
```
Isn't it too complex?

Reply:
I think we should create a base model that can represent all the event types, except the keep alive type, ie a model that includes the `ha_id`.

If there are event types where we can't use the current `Event` (I assume this since you check for `EventType.STATUS | EventType.EVENT | EventType.NOTIFY`, before using this model), we can create different child model classes that inherit the base model. One would be the current `Event` model (but it should inherit the new base model), and the other(s) would represent the remaining event types as needed (CONNECTED, DISCONNECTED, PAIRED, DEPAIRED).

We should not create more models than needed to represent the different event types. Ie one model can represent more than one event type. It depends on what the possible model attributes are for the different event types. If an event type can never return an attribute of a model, that model isn't suitable to represent that event type, but another model is needed. Optional attributes that are present some of the time for an event type is of course ok to have in the model.

What attributes do the CONNECTED, DISCONNECTED, PAIRED, DEPAIRED event types return?

Reply:
After experimenting for a while, this is what each type of event sends:
NOTIFY:
```json
{
  "haId": "SIEMENS-HCS02DWH1-6BE58C26DCC1",
  "items": [
    {
      "handling": "none",
      "key": "Dishcare.Dishwasher.Option.HalfLoad",
      "level": "hint",
      "timestamp": 1733611453,
      "uri": "/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DCC1/programs/selected/options/Dishcare.Dishwasher.Option.HalfLoad",
      "value": true
    },
    {
      "handling": "none",
      "key": "BSH.Common.Option.RemainingProgramTime",
      "level": "hint",
      "timestamp": 1733611453,
      "unit": "seconds",
      "uri": "/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DCC1/programs/selected/options/BSH.Common.Option.RemainingProgramTime",
      "value": 13800
    }
  ]
}
```

EVENT:
```json
{
  "haId": "SIEMENS-HCS02DWH1-6BE58C26DCC1",
  "items": [
    {
      "handling": "acknowledge",
      "key": "BSH.Common.Event.ProgramAborted",
      "level": "hint",
      "timestamp": 1733616930,
      "value": "BSH.Common.EnumType.EventPresentState.Present"
    }
  ]
}
```

STATUS:
```json
{
  "haId": "SIEMENS-HCS02DWH1-6BE58C26DCC1",
  "items": [
    {
      "handling": "none",
      "key": "BSH.Common.Status.RemoteControlStartAllowed",
      "level": "hint",
      "timestamp": 1733611714,
      "uri": "/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DCC1/status/BSH.Common.Status.RemoteControlStartAllowed",
      "value": true
    }
  ]
}
```

DISCONNECTED (and probably CONNECTED, PAIRED, DEPAIRED too):
```json
{
  "haId": "SIEMENS-HCS02DWH1-6BE58C26DCC1",
  "handling": "none",
  "key": "BSH.Common.Appliance.Disconnected",
  "level": "hint",
  "timestamp": 1733611817,
  "value": true
}
```

So, in conclusion:

- NOTIFY and STATUS apparently uses uri, EVENT seems to not use it (as you cannot fetch events)
- `unit` value depends completely on the key sent
- After using `de-DE` as accepted language `displayValue` was always set
- PAIRED, DEPAIRED, DEPAIRED and DISCONNECTED sends just one event, not even in a `items` array
- Home Appliance ID is always sent, so we will need to add it to `ArrayOfEvents` and `Event` data classes with (at `Event` data class, the field should be defaulted to `None`)

If we model based on this responses, we would have the following classes:
- Base event:
  - handling
  - key
  - level
  - timestamp
  - value
- Event for STATUS and NOTIFY event types, which inherits from base event and adds:
  - name, which defaults to `None`
  - display value, which defaults to `None`
  - uri, which defaults to `None`
  - unit, which defaults to `None`
- Event for PAIRED, DEPAIRED, DEPAIRED and DISCONNECTED event types , which inherits from base event and adds:
  -  home appliance ID

I added the name, display value and unit fields defaulted to `None` because API documentation specifies that they are optional fields, so they might or might not be sent. And because of that, in my opinion, I think we shouldn't try to model based on the obtained responses and create the model strictly based on the model given at API documentation: A Event data class with optional fields (So maintain what we have right now).
In the end, the API model is what we can be pretty sure that would work at any situation.

The only thing I would add, is the event class for PAIRED, DEPAIRED, DEPAIRED and DISCONNECTED event types, which would be at an union with `ArrayOfEvents` at the message model that I'm preparing:
```python
@dataclass
class EventMessage:
    id: str
    type: EventType
    data: ArrayOfEvents | EventWithHaId | None = None
```
(`None` is for KEEP-ALIVE event type, which doesn't contain any data)

Reply:
Doesn't all the events contain `ha_id` except the keep alive event type? It looks like that in your examples.

Reply:
I think we should consider flattening the data so that all events have a similar structure and only one level.

Reply:
>Doesn't all the events contain ha_id except the keep alive event type? It looks like that in your examples.

Yep, All the JSONs posted are the raw data extracted from the SSE events data.

Reply:
> I think we should consider flattening the data so that all events have a similar structure and only one level. Ie create one event for each item in `items`.

That's what we already do, right?

Reply:
I was thinking something like this:

```py
@dataclass
class EventMessage:
    """Represent a server sent event message sent from the Home Connect API."""

    id: str
    type: EventType
    data: Event

    @classmethod
    def from_server_sent_event(cls, sse: ServerSentEvent, event: Event) -> Self:   
        return cls(
            id=sse.id,
            type=event_type = EventType(sse.event),
            data=event,
        )

async def stream_events(...) -> AsyncGenerator[EventMessage]:
    ...
    async for sse in event_source.aiter_sse(): 
        data = sse.data
        if isinstance(data, list):
            for item in data
                event = Event.from_json(item)
                yield EventMessage.from_server_sent_event(sse, event)
        else:
            event = Event.from_json(data)
            yield EventMessage.from_server_sent_event(sse, event)    
```

Reply:
Oooh, I see!
Seems a pretty good idea to me!

Reply:
About the `haId` property that comes at SSE's data, although `haId` is usually equal to the value of the SSE event's `id`, should we add it to se `Event` data class?
Or maybe should we do something similar to this?
```
id = sse.data.get("haId") or sse.id
```
(I would prefer the first one to make sure we don't loose data)

Reply:
We can add it, ie your first suggestion.